### PR TITLE
scenario tests

### PR DESCRIPTION
**Pull request by PietroPasotti**
This PR adds scenario tests to traefik.

### CODE REVIEW

**Comment by Abuelodelanada** [comment]
To run these test you need to install ops-scenario from [this wheel](https://github.com/PietroPasotti/ops-scenario/releases/tag/0.2.2) and then:

```shell
$ tox -e scenario -- tests/scenario/test_setup.py
scenario: commands[0]> pytest -v --tb native --ignore=/home/jose/trabajos/canonical/repos/traefik-k8s-operator/tests/unit --ignore=/home/jose/trabajos/canonical/repos/traefik-k8s-operator/tests/integration --log-cli-level=INFO -s tests/scenario/test_setup.py
======================================================= test session starts ========================================================
platform linux -- Python 3.10.7, pytest-7.1.2, pluggy-1.0.0 -- /usr/bin/python3
cachedir: .tox/scenario/.pytest_cache
rootdir: /home/jose/trabajos/canonical/repos/traefik-k8s-operator, configfile: pyproject.toml
plugins: dependency-0.5.1, anyio-3.6.2, cov-3.0.0
collected 4 items                                                                                                                  

tests/scenario/test_setup.py::test_start_traefik_is_not_running PASSED
tests/scenario/test_setup.py::test_start_traefik_no_hostname PASSED
tests/scenario/test_setup.py::test_start_traefik_active PASSED
tests/scenario/test_setup.py::test_start_traefik_invalid_routing_mode 
---------------------------------------------------------- live log call -----------------------------------------------------------
ERROR    charm:charm.py:265 'invalid_routing' is not a valid routing_mode value; valid values are: ['path', 'subdomain']
PASSED

======================================================== 4 passed in 1.20s =========================================================
  scenario: OK (2.15=setup[0.10]+cmd[2.04] seconds)
  congratulations :) (2.28 seconds)
```

**Comment by sed-i** [comment]
After changing
```diff
 deps =
+    pytest
     -r{toxinidir}/requirements.txt
+    git+https://github.com/PietroPasotti/ops-scenario
```
I tried `tox -e scenario` and got
```
tests/scenario/test_lib_per_unit_provides_sequences.py:6: in <module>
    from scenario import Scenario, Scene
E   ImportError: cannot import name 'Scenario' from 'scenario' (unknown location)
```
An alternative import works though:
```
$ source .tox/scenario/bin/activate
$ python
>>> from scenario.scenario import Scenario
>>> 
```
What am I doing wrong?

**Comment by sed-i** [comment]
Same problem with the wheel:
```diff
 deps =
+    pytest
     -r{toxinidir}/requirements.txt
+    https://github.com/PietroPasotti/ops-scenario/releases/download/0.2.2/scenario-0.2.2-py3-none-any.whl
```
```
$ tox -re scenario
tests/scenario/test_lib_per_unit_provides_sequences.py:6: in <module>
    from scenario import Scenario, Scene
E   ImportError: cannot import name 'Scenario' from 'scenario' (unknown location)
```

**Comment by Abuelodelanada** [comment]
I did it manually:

```sh
pip install scenario-0.2.2-py3-none-any.whl
``` 

and then run:

```sh
tox -e scenario
```

**Comment by PietroPasotti** [comment]
@Abuelodelanada @sed-i FYI I just pushed a tox.init change that pins the scenario dependency. No need to manually install the wheel if you're running `tox -e scenario`.

If you add this line to your `requirements.txt` it should do the right thing on `pip install -r`:
`git+https://github.com/PietroPasotti/ops-scenario@1.0.0#egg=scenario`

**Comment by sed-i** [comment]
Why CI passes? Scenario tests fail locally.

File: .github/workflows/pull-request.yaml [hunk 1]
```diff
@@ -10,3 +10,22 @@ jobs:
...
+      - static-checks
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
+      - name: Install charmcraft
+        run: snap install charmcraft
+      - name: Fetch dependencies
+        run: charmcraft fetch-lib charms.harness_extensions.v0.evt_sequences
```
Comment (sed-i) [inline .github/workflows/pull-request.yaml:29]: I guess we can drop this one now?

File: pyproject.toml [hunk 2]
```diff
@@ -11,7 +11,6 @@ show_missing = true
 [tool.pytest.ini_options]
 minversion = "6.0"
 log_cli_level = "INFO"
-asyncio_mode = "auto"
```
Comment (sed-i) [inline pyproject.toml:14]: Afaik this one is currently necessary for itests to work

**Comment by sed-i** [comment]
@PietroPasotti after `tox -re scenario` I am still getting
```
tests/scenario/test_setup.py:8: in <module>
    from scenario.scenario import Scenario
E   ModuleNotFoundError: No module named 'scenario.scenario'
```

**Comment by PietroPasotti** [comment]
> @PietroPasotti after `tox -re scenario` I am still getting
> 
> ```
> tests/scenario/test_setup.py:8: in <module>
>     from scenario.scenario import Scenario
> E   ModuleNotFoundError: No module named 'scenario.scenario'
> ```

Fixed.

File: tests/scenario/test_status.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
+import unittest
```
Comment (sed-i) [inline tests/scenario/test_status.py:1]: Note: this file is a 1:1 rewrite of 'test_setup'

**Comment by PietroPasotti** [comment]
Are we ever merging this PR or is it just a sandbox for y'all to take turns playing with scenario?

**Comment by sed-i** [comment]
> Are we ever merging this PR or is it just a sandbox for y'all to take turns playing with scenario?

The PR isn't ready. At the moment, test_lib... doesn't assert anything.

But we should probably state the scope of the PR in the description :P

File: tox.ini [hunk 4]
```diff
@@ -4,7 +4,7 @@
 [tox]
 skipsdist=True
 skip_missing_interpreters = True
-envlist = lint, static-{charm, lib}, unit
+envlist = lint, static-{charm, lib}, unit, scenario
 
 [vars]
 src_path = {toxinidir}/src/
```
Comment (sed-i) [inline tox.ini:10]: By convention these vars do not have a trailing `/` and when used below we do use a `/`.

**Review by lucabello — APPROVED** [review]

**Review by rbarry82 — APPROVED** [review]

**Pull request merged**